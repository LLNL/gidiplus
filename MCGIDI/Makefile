SHELL = /bin/ksh

# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>

DIRS = Src include lib bin Doc

POPs_PATH = $(abspath ../PoPI)
SMR_PATH = $(abspath ../statusMessageReporting)
NF_PATH = $(abspath ../numericalFunctions)

.PHONY: default include lib bin clean realclean

default: include lib
	cd Test; $(MAKE) default
	$(MAKE) bin

include:
	cd include; $(MAKE)

lib:
	cd lib; $(MAKE) CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' POPs_PATH='$(POPs_PATH)' SMR_PATH='$(SMR_PATH)' NF_PATH='$(NF_PATH)'

bin:
	cd bin; $(MAKE) CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' POPs_PATH='$(POPs_PATH)' SMR_PATH='$(SMR_PATH)' NF_PATH='$(NF_PATH)' bin

check:
	cd Test; $(MAKE) CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' POPs_PATH='$(POPs_PATH)' SMR_PATH='$(SMR_PATH)' NF_PATH='$(NF_PATH)' check

speeds:
	cd Speeds; $(MAKE) CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' POPs_PATH='$(POPs_PATH)' SMR_PATH='$(SMR_PATH)' NF_PATH='$(NF_PATH)' speeds

clean:
	$(MAKE) doDIRS _DIRS="$(DIRS)" _TARGET=clean
	cd Test; $(MAKE) clean
	cd Speeds; $(MAKE) clean

realclean:
	$(MAKE) doDIRS _DIRS="$(DIRS)" _TARGET=realclean
	cd Test; $(MAKE) realclean
	cd Speeds; $(MAKE) realclean

docs:
	doxygen Doc/doxygen/Doxyfile

doDIRS:
	SAVED_PWD=`pwd`; \
	for DIR in $(_DIRS); do cd $$DIR; $(MAKE) $(_TARGET); cd $$SAVED_PWD; done
