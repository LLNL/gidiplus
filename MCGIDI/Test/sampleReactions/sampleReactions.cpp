/*
# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>
*/

#include <stdlib.h>
#include <iostream>
#include <set>

#include "MCGIDI.hpp"

#include "MCGIDI_testUtilities.hpp"

static char const *description = "Loops over energy at the specified temperature, sampling reactions. If projectile is a photon, see options *-pa* and *-pn*.";

void main2( int argc, char **argv );
/*
=========================================================
*/
int main( int argc, char **argv ) {

    try {
        main2( argc, argv );
        exit( EXIT_SUCCESS ); }
    catch (std::exception &exception) {
        std::cerr << exception.what( ) << std::endl; }
    catch (char const *str) {
        std::cout << str << std::endl; }
    catch (std::string &str) {
        std::cout << str << std::endl;
    }
}
/*
=========================================================
*/
void main2( int argc, char **argv ) {

    PoPI::Database pops( "../../../TestData/PoPs/pops.xml" );
    GIDI::Transporting::Particles particles;
    char *endChar;
    std::size_t numberOfSamples = 1000 * 1000;
    void *rngState = nullptr;
    unsigned long long seed = 1;
    std::set<int> reactionsToExclude;
    LUPI::StatusMessageReporting smr1;
    GIDI::Construction::PhotoMode photo_mode = GIDI::Construction::PhotoMode::nuclearOnly;

    std::cerr << "    " << __FILE__;
    for( int i1 = 1; i1 < argc; i1++ ) std::cerr << " " << argv[i1];
    std::cerr << std::endl;

    MCGIDI_test_rngSetup( seed );

    argvOptions2 argv_options( "sampleReactions", description );

    argv_options.add( argvOption2( "--pid", true, "The PoPs id of the projectile." ) );
    argv_options.add( argvOption2( "--tid", true, "The PoPs id of the target." ) );
    argv_options.add( argvOption2( "--map", true, "The map file to use." ) );
    argv_options.add( argvOption2( "-pa", false, "Include photo-atomic protare if relevant. If present, disables photo-nuclear unless *-pn* present." ) );
    argv_options.add( argvOption2( "-pn", false, "Include photo-nuclear protare if relevant. This is the default unless *-pa* present." ) );
    argv_options.add( argvOption2( "--fixedGrid", false, "Set fixed grid data. Only used if protare is only photo-atomic protare." ) );
    argv_options.add( argvOption2( "--temperature", true, "The temperature of the target material." ) );

    argv_options.parseArgv( argc, argv );

    std::string projectileID = argv_options.find( "--pid" )->zeroOrOneOption( argv, PoPI::IDs::neutron );
    std::string targetID = argv_options.find( "--tid" )->zeroOrOneOption( argv, "O16" );
    std::string mapFilename = argv_options.find( "--map" )->zeroOrOneOption( argv, "../../../GIDI/Test/all3T.map" );

    if( argv_options.find( "-pa" )->present( ) ) {
        photo_mode = GIDI::Construction::PhotoMode::atomicOnly;
        if( argv_options.find( "-pn" )->present( ) ) photo_mode = GIDI::Construction::PhotoMode::nuclearAndAtomic;
    }

    GIDI::Map::Map map( mapFilename, pops );

    std::string temperatureString = argv_options.find( "--temperature" )->zeroOrOneOption( argv, "2.58522e-8" );
    double temperature = strtod( temperatureString.c_str( ), &endChar );
    if( *endChar != 0 ) throw "Invalid temperature input.";

    GIDI::Construction::Settings construction( GIDI::Construction::ParseMode::all, photo_mode );
    GIDI::Protare *protare = map.protare( construction, pops, projectileID, targetID );

    GIDI::Styles::TemperatureInfos temperatures = protare->temperatures( );
    for( GIDI::Styles::TemperatureInfos::const_iterator iter = temperatures.begin( ); iter != temperatures.end( ); ++iter ) {
        std::cout << "label = " << iter->heatedCrossSection( ) << "  temperature = " << iter->temperature( ).value( ) << std::endl;
    }

    std::string label( temperatures[0].griddedCrossSection( ) );
    MCGIDI::Transporting::MC MC( pops, projectileID, &protare->styles( ), label, GIDI::Transporting::DelayedNeutrons::on, 20.0 );

    if( argv_options.find( "--fixedGrid" )->present( ) ) {
        GIDI::Groups groups( "../../../GIDI/Test/groups.xml" );
        MC.fixedGridPoints( groups.get<GIDI::Group>( "LLNL_gid_80" )->data( ) );
    }

    MCGIDI::DomainHash domainHash( 4000, 1e-8, 10 );
    MCGIDI::Protare *MCProtare = MCGIDI::protareFromGIDIProtare( smr1, *protare, pops, MC, particles, domainHash, temperatures, reactionsToExclude );

    MCGIDI::Vector<MCGIDI::Protare *> protares( 1 );
    protares[0] = MCProtare;
    MCGIDI::URR_protareInfos URR_protare_infos( protares );

    int numberOfReactions = (int) MCProtare->numberOfReactions( );

    if( argv_options.find( "--fixedGrid" )->present( ) ) {
        for( std::size_t reactionIndex = 0; reactionIndex < MCProtare->numberOfReactions( ); ++reactionIndex ) {
            MCGIDI::Reaction const *reaction = MCProtare->reaction( reactionIndex );

            std::cout << reaction->label( ).c_str( ) << std::endl;
        }
    }

    std::cout << "temperature = " << temperature << std::endl;
    std::cout << "numberOfSamples = " << numberOfSamples << std::endl;
    for( double energy = 1e-12; energy < 25.0; energy *= 1.2 ) {
        int hashIndex = domainHash.index( energy );

        double crossSection = MCProtare->crossSection( URR_protare_infos, hashIndex, temperature, energy );
        std::cout << "energy = " << energy << " " << "cross section = " << doubleToString2( "%13.6e", crossSection ) << std::endl;
        if( crossSection == 0.0 ) continue;

        std::vector<double> reactionCrossSections( numberOfReactions );
        std::cout << "      ";
        for( int i1 = 0; i1 < numberOfReactions; ++i1 ) {
            double reactionCrossSection = MCProtare->reactionCrossSection( i1, URR_protare_infos, hashIndex, temperature, energy );
            std::cout << LUPI::Misc::argumentsToString( " %9.6f", reactionCrossSection / crossSection );
        }
        std::cout << std::endl;

        std::vector<long> counts( numberOfReactions + 1, 0 );
        for( std::size_t i1 = 0; i1 < numberOfSamples; ++i1 ) {
            int reactionIndex = MCProtare->sampleReaction( URR_protare_infos, hashIndex, temperature, energy, crossSection, float64RNG64, rngState );
            if( reactionIndex > numberOfReactions ) reactionIndex = numberOfReactions;
            ++counts[reactionIndex];
        }

        std::cout << "      ";
        for( int i1 = 0; i1 < numberOfReactions; ++i1 ) {
            double ratio = counts[i1];
            std::cout << LUPI::Misc::argumentsToString( " %9.6f", ratio / numberOfSamples );
        }
        std::cout << std::endl;

        std::cout << "      ";
        for( int i1 = 0; i1 < numberOfReactions; ++i1 ) {
            std::cout << LUPI::Misc::argumentsToString( " %9ld", counts[i1] );
        }
        std::cout << std::endl;
    }

    delete protare;

    delete MCProtare;
}
