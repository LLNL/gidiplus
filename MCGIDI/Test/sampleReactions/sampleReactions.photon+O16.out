label = heated_000  temperature = 0
temperature = 2.58522e-08
numberOfSamples = 1000000
energy = 1e-12 cross section =  0.000000e+00
energy = 1.2e-12 cross section =  0.000000e+00
energy = 1.44e-12 cross section =  0.000000e+00
energy = 1.728e-12 cross section =  0.000000e+00
energy = 2.0736e-12 cross section =  0.000000e+00
energy = 2.48832e-12 cross section =  0.000000e+00
energy = 2.98598e-12 cross section =  0.000000e+00
energy = 3.58318e-12 cross section =  0.000000e+00
energy = 4.29982e-12 cross section =  0.000000e+00
energy = 5.15978e-12 cross section =  0.000000e+00
energy = 6.19174e-12 cross section =  0.000000e+00
energy = 7.43008e-12 cross section =  0.000000e+00
energy = 8.9161e-12 cross section =  0.000000e+00
energy = 1.06993e-11 cross section =  0.000000e+00
energy = 1.28392e-11 cross section =  0.000000e+00
energy = 1.5407e-11 cross section =  0.000000e+00
energy = 1.84884e-11 cross section =  0.000000e+00
energy = 2.21861e-11 cross section =  0.000000e+00
energy = 2.66233e-11 cross section =  0.000000e+00
energy = 3.1948e-11 cross section =  0.000000e+00
energy = 3.83376e-11 cross section =  0.000000e+00
energy = 4.60051e-11 cross section =  0.000000e+00
energy = 5.52061e-11 cross section =  0.000000e+00
energy = 6.62474e-11 cross section =  0.000000e+00
energy = 7.94968e-11 cross section =  0.000000e+00
energy = 9.53962e-11 cross section =  0.000000e+00
energy = 1.14475e-10 cross section =  0.000000e+00
energy = 1.37371e-10 cross section =  0.000000e+00
energy = 1.64845e-10 cross section =  0.000000e+00
energy = 1.97814e-10 cross section =  0.000000e+00
energy = 2.37376e-10 cross section =  0.000000e+00
energy = 2.84852e-10 cross section =  0.000000e+00
energy = 3.41822e-10 cross section =  0.000000e+00
energy = 4.10186e-10 cross section =  0.000000e+00
energy = 4.92224e-10 cross section =  0.000000e+00
energy = 5.90668e-10 cross section =  0.000000e+00
energy = 7.08802e-10 cross section =  0.000000e+00
energy = 8.50562e-10 cross section =  0.000000e+00
energy = 1.02067e-09 cross section =  0.000000e+00
energy = 1.22481e-09 cross section =  0.000000e+00
energy = 1.46977e-09 cross section =  0.000000e+00
energy = 1.76373e-09 cross section =  0.000000e+00
energy = 2.11647e-09 cross section =  0.000000e+00
energy = 2.53977e-09 cross section =  0.000000e+00
energy = 3.04772e-09 cross section =  0.000000e+00
energy = 3.65726e-09 cross section =  0.000000e+00
energy = 4.38871e-09 cross section =  0.000000e+00
energy = 5.26646e-09 cross section =  0.000000e+00
energy = 6.31975e-09 cross section =  0.000000e+00
energy = 7.5837e-09 cross section =  0.000000e+00
energy = 9.10044e-09 cross section =  0.000000e+00
energy = 1.09205e-08 cross section =  0.000000e+00
energy = 1.31046e-08 cross section =  0.000000e+00
energy = 1.57256e-08 cross section =  0.000000e+00
energy = 1.88707e-08 cross section =  0.000000e+00
energy = 2.26448e-08 cross section =  0.000000e+00
energy = 2.71738e-08 cross section =  0.000000e+00
energy = 3.26085e-08 cross section =  0.000000e+00
energy = 3.91302e-08 cross section =  0.000000e+00
energy = 4.69563e-08 cross section =  0.000000e+00
energy = 5.63475e-08 cross section =  0.000000e+00
energy = 6.7617e-08 cross section =  0.000000e+00
energy = 8.11404e-08 cross section =  0.000000e+00
energy = 9.73685e-08 cross section =  0.000000e+00
energy = 1.16842e-07 cross section =  0.000000e+00
energy = 1.40211e-07 cross section =  0.000000e+00
energy = 1.68253e-07 cross section =  0.000000e+00
energy = 2.01903e-07 cross section =  0.000000e+00
energy = 2.42284e-07 cross section =  0.000000e+00
energy = 2.90741e-07 cross section =  0.000000e+00
energy = 3.48889e-07 cross section =  0.000000e+00
energy = 4.18667e-07 cross section =  0.000000e+00
energy = 5.024e-07 cross section =  0.000000e+00
energy = 6.0288e-07 cross section =  0.000000e+00
energy = 7.23456e-07 cross section =  0.000000e+00
energy = 8.68147e-07 cross section =  0.000000e+00
energy = 1.04178e-06 cross section =  0.000000e+00
energy = 1.25013e-06 cross section =  0.000000e+00
energy = 1.50016e-06 cross section =  0.000000e+00
energy = 1.80019e-06 cross section =  0.000000e+00
energy = 2.16023e-06 cross section =  0.000000e+00
energy = 2.59227e-06 cross section =  0.000000e+00
energy = 3.11073e-06 cross section =  0.000000e+00
energy = 3.73287e-06 cross section =  0.000000e+00
energy = 4.47945e-06 cross section =  0.000000e+00
energy = 5.37534e-06 cross section =  0.000000e+00
energy = 6.45041e-06 cross section =  0.000000e+00
energy = 7.74049e-06 cross section =  0.000000e+00
energy = 9.28859e-06 cross section =  0.000000e+00
energy = 1.11463e-05 cross section =  0.000000e+00
energy = 1.33756e-05 cross section =  0.000000e+00
energy = 1.60507e-05 cross section =  0.000000e+00
energy = 1.92608e-05 cross section =  0.000000e+00
energy = 2.3113e-05 cross section =  0.000000e+00
energy = 2.77356e-05 cross section =  0.000000e+00
energy = 3.32827e-05 cross section =  0.000000e+00
energy = 3.99392e-05 cross section =  0.000000e+00
energy = 4.79271e-05 cross section =  0.000000e+00
energy = 5.75125e-05 cross section =  0.000000e+00
energy = 6.9015e-05 cross section =  0.000000e+00
energy = 8.2818e-05 cross section =  0.000000e+00
energy = 9.93816e-05 cross section =  0.000000e+00
energy = 0.000119258 cross section =  0.000000e+00
energy = 0.000143109 cross section =  0.000000e+00
energy = 0.000171731 cross section =  0.000000e+00
energy = 0.000206078 cross section =  0.000000e+00
energy = 0.000247293 cross section =  0.000000e+00
energy = 0.000296752 cross section =  0.000000e+00
energy = 0.000356102 cross section =  0.000000e+00
energy = 0.000427323 cross section =  0.000000e+00
energy = 0.000512787 cross section =  0.000000e+00
energy = 0.000615344 cross section =  0.000000e+00
energy = 0.000738413 cross section =  0.000000e+00
energy = 0.000886096 cross section =  0.000000e+00
energy = 0.00106332 cross section =  0.000000e+00
energy = 0.00127598 cross section =  0.000000e+00
energy = 0.00153117 cross section =  0.000000e+00
energy = 0.00183741 cross section =  0.000000e+00
energy = 0.00220489 cross section =  0.000000e+00
energy = 0.00264587 cross section =  0.000000e+00
energy = 0.00317504 cross section =  0.000000e+00
energy = 0.00381005 cross section =  0.000000e+00
energy = 0.00457206 cross section =  0.000000e+00
energy = 0.00548647 cross section =  0.000000e+00
energy = 0.00658377 cross section =  0.000000e+00
energy = 0.00790052 cross section =  0.000000e+00
energy = 0.00948063 cross section =  0.000000e+00
energy = 0.0113768 cross section =  0.000000e+00
energy = 0.0136521 cross section =  0.000000e+00
energy = 0.0163825 cross section =  0.000000e+00
energy = 0.019659 cross section =  0.000000e+00
energy = 0.0235908 cross section =  0.000000e+00
energy = 0.028309 cross section =  0.000000e+00
energy = 0.0339708 cross section =  0.000000e+00
energy = 0.040765 cross section =  0.000000e+00
energy = 0.0489179 cross section =  0.000000e+00
energy = 0.0587015 cross section =  0.000000e+00
energy = 0.0704418 cross section =  0.000000e+00
energy = 0.0845302 cross section =  0.000000e+00
energy = 0.101436 cross section =  0.000000e+00
energy = 0.121724 cross section =  0.000000e+00
energy = 0.146068 cross section =  0.000000e+00
energy = 0.175282 cross section =  0.000000e+00
energy = 0.210338 cross section =  0.000000e+00
energy = 0.252406 cross section =  0.000000e+00
energy = 0.302887 cross section =  0.000000e+00
energy = 0.363464 cross section =  0.000000e+00
energy = 0.436157 cross section =  0.000000e+00
energy = 0.523389 cross section =  0.000000e+00
energy = 0.628067 cross section =  0.000000e+00
energy = 0.75368 cross section =  0.000000e+00
energy = 0.904416 cross section =  0.000000e+00
energy = 1.0853 cross section =  0.000000e+00
energy = 1.30236 cross section =  0.000000e+00
energy = 1.56283 cross section =  0.000000e+00
energy = 1.8754 cross section =  0.000000e+00
energy = 2.25048 cross section =  0.000000e+00
energy = 2.70057 cross section =  0.000000e+00
energy = 3.24069 cross section =  0.000000e+00
energy = 3.88882 cross section =  0.000000e+00
energy = 4.66659 cross section =  0.000000e+00
energy = 5.5999 cross section =  0.000000e+00
energy = 6.71989 cross section =  0.000000e+00
energy = 8.06386 cross section =  0.000000e+00
energy = 9.67663 cross section =  0.000000e+00
energy = 11.612 cross section =  0.000000e+00
energy = 13.9344 cross section =  6.969376e-04
        0.000000  1.000000
        0.000000  1.000000
               0   1000000
energy = 16.7212 cross section =  4.205069e-03
        0.105218  0.894782
        0.105318  0.894682
          105318    894682
energy = 20.0655 cross section =  2.316051e-03
        0.304290  0.695710
        0.304016  0.695984
          304016    695984
energy = 24.0786 cross section =  1.206275e-02
        0.461634  0.538366
        0.461896  0.538104
          461896    538104
